File: 108-convert-sorted-array-to-binary-search-tree/108-convert-sorted-array-to-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {

     private TreeNode binaryBuilder(int[] nums, int left, int right){
        if (left > right) return null;
         
        int mid = (right + left) / 2;
         
        TreeNode root = new TreeNode(nums[mid]);
         
        root.left = binaryBuilder(nums, left, mid - 1);
        root.right = binaryBuilder(nums, mid + 1, right);
         
        return root;
    }
    
    public TreeNode sortedArrayToBST(int[] nums) {
        return binaryBuilder(nums,0, nums.length-1) ;
    }
}

File: 108-convert-sorted-array-to-binary-search-tree/108-convert-sorted-array-to-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode *helper(vector<int>&nums, int start, int end){
        if(start>end) return NULL;
        int mid = (start+end)/2;
        TreeNode *newNode = new TreeNode(nums[mid]);
        newNode->left =  helper(nums, start, mid-1);
        newNode->right = helper(nums, mid+1, end);
        return newNode;
    }
    TreeNode* sortedArrayToBST(vector<int>& nums) {
        int n = nums.size();
        return helper(nums, 0, n-1);
    }
};